File: checks/python/src/section508/checks/focus_order.py
from section508.dom import attr, parse_number, select


def check(root):
    if not root:
        return []

    positive = []
    for element in select(root, "[tabindex]"):
        value = parse_number(attr(element, "tabindex"))
        if value is not None and value > 0:
            positive.append(element)
    if positive:
        return ["Positive tabindex values reorder Tab away from the DOM."]
    return []

File: checks/ruby/lib/section508/checks/focus_order.rb
# frozen_string_literal: true

module Section508
  module Checks
    # Positive tabindex values reorder Tab away from the DOM.
    class FocusOrder
      def self.call(root)
        return [] unless root&.respond_to?(:css)

        positive = Dom.elements(root, "[tabindex]").select do |element|
          value = number(element["tabindex"])
          value.finite? && value > 0
        end
        return ["Positive tabindex values reorder Tab away from the DOM."] if positive.any?

        []
      end

      def self.number(value)
        Float(value)
      rescue ArgumentError, TypeError
        Float::NAN
      end
    end
  end
end
